### ISSUE TITLE
stripping em units is not very useful

### ISSUE THREAD

**Issue by heygrady** [DriveWealth/gulp-react-native-stylesheet-css#2]
When a length value is encountered ([here](https://github.com/DriveWealth/gulp-react-native-stylesheet-css/blob/master/lib/properties/all.js#L16) and [here](https://github.com/DriveWealth/gulp-react-native-stylesheet-css/blob/master/lib/properties/all.js#L25)), px and em are automatically stripped. This isn't the best way to do it and produces undesired results. Specifically, most units are preserved and em units are treated as being the same as pixels. In practice, `1em = 16px` instead of `1em = 1px`.

**input.css**
```css
container: {
  font-size: 16px;
  width: 1em;
  height: 1rem;
}
```

**output.js**
```js
var React = require('react-native');
var { StyleSheet } = React;
module.exports = StyleSheet.create({
    "container": {
        "fontSize": 16,
        "width": 1,
        "height": "1rem"
    }
});
```

In this case the unit stripping beyond `px` seems arbitrary and doesn't produce useful results. I worked on a [unit converter](https://github.com/heygrady/Units#length) (only works in browsers) a few years ago. There are a bunch of standard CSS units that convert easily to px and the relative units like `em` can usually be assumed to be relative to 16px better than presuming 1px.

Units like `pt` can be cleanly converted to `px` (`72pt === 96px`). Other units like `vh` can actually be calculated in React Native if we output something better in the stylesheet.

### Example of using React Native for `vh` units:

**input.css**
```css
container: {
  height: 2vh;
}
```

**output.js**
```js
var React = require('react-native');
var { StyleSheet, Dimensions } = React;
var viewport = Dimensions.get('window');

function vh(val) {
  return viewport.height / 100 * val;
}

module.exports = StyleSheet.create({
    "container": {
        "height": vh(2)
    }
});
```

### Example of using React Native for `em` units:

**input.css**
```css
container: {
  height: 1em;
}
```

**output.js**
```js
var React = require('react-native');
var { StyleSheet } = React;

var parentFontSize = 16; // or maybe require('./variables').parentFontSize;

function em(val) {
  return parentFontSize * val;
}

module.exports = StyleSheet.create({
    "container": {
        "height": em(2)
    }
});
```

### What I wanted to do
I wanted to write a gulp plugin that injected something like `var { em, vh, percentage } = require('my-projects-units-utility');` at the top of my outputted files and converted values like `1em` to `em(1)` so that the unit library could handle it appropriately. Of course this is most useful for importing CSS from non-React-Native projects.

It would be possible to do something like this in a component:

**input.css**
```css
container: {
  height: 2em;
}
```

**output.js**
```js
var React = require('react-native');
var { StyleSheet } = React;
var { em } = require('my-projects-units-utility');

module.exports = StyleSheet.create({
    "container": {
        "height": em(2, 'container')
    }
});
```

**component.js**
```js
import React, { Component, View } from 'react-native';
import { configSelector } from 'my-projects-units-utility';

// set the relative unit for a given selector.
// defaults would be something reasonable like 16px for parentFontSize and baseFontSize
//   or viewport.width for parentWidth
configSelector('container', { parentFontSize: 12 });

import styles from 'output';

export default class MyThing extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <View style={styles.container}>{this.props.children}</View>
    );
  }
}
```

**Comment by jjshammas** [comment]
Hi @heygrady:

Thank you very much for your very thorough comments. You're absolutely right that, semantically, it doesn't make much sense for `1em` to wind up being interpreted as `1px`.

What we wanted to accomplish in this project was to keep the styles separated from JavaScript, more inline with how web projects are built. This allows use to leverage the immense power of SASS for things like color operations. One of the downsides to this separation is—at least with RN's StyleSheet object—there's no proper nesting and calculating the proper size for an `em` wouldn't be possible.

I definitely think it's worth exploring something that can handle nesting more akin to how real CSS works—likely driven by JavaScript. That would wind up being a project separate from this one.

You might want to check out [react-native-extended-stylesheet](https://github.com/vitalets/react-native-extended-stylesheet) which simulates a lot of CSS rules, albeit requiring you to write all of your style declarations in JavaScript.

**Comment by heygrady** [comment]
Stripping em but leaving units like rem and ch is frustrating. Using gulp I would be able to doctor the conversions as I see fit for my project with my own plugin jammed into the pipeline... except em values are stripped.

It would be nice to be able to supply a list of defaults, like 1em = 16px, as options to your plugin, or be able to convert them on my own in my own plugin.

Point taken on "nesting doesn't work." I had a plan to make something that works for me but unfortunately you're stripping em units without converting them to pixels first.

**Comment by jjshammas** [comment]
@heygrady you're right about the inconsistency. I think the first step will either be stripping all units, or none at all.

As far as passing a value for `em`s, I think that doesn't make sense for `em` but would make sense for `rem`. Since `em`s are dependent on the parent element's font size in CSS, allowing a developer to pass in a single value could cause confusion. Adding such support for units not based on the flow of a document (like `rem`) would be a nice potential improvement.

**Comment by heygrady** [comment]
Huge vote for leaving unsupported or unconvertible units in place. This would allow for novel use-cases like what I'm describing. For instance, `em` is totally knowable when a component is being rendered. Stripping units without converting them will stop someone from, say, creating a gulp plugin from EStyleSheet and using your awesome plugin as a parser/converter to feed CSS styles into the JS pipeline.

**Fictional example of what might be done by a gulp plugin using the output of `gulp-react-native-stylesheet-css` if it didn't strip unknown/unsupported units**
```js
import React, { Component, View, Text } from 'react-native';

// special StyleSheet object that supports creating with context
import { StyleSheet } from 'my-special-stylesheet';

// special output from a project-specific-gulp-plugin
// alters the output from gulp-react-native-stylesheet-css
  // to support customizations for units
  // or other things, like media-queries
  // or otherwise customize the output of the styles.js file
  // to suit the needs of the project
import { rawStyles } from './styles/output';

export default class MyThing extends Component {
  constructor(props) {
    super(props);
  }
  
  // allow parent/root contextual styles to be passed in from the outside
  static propTypes = {
    parentViewStyle: View.propTypes.style,
    parentTextStyle: Text.propTypes.style,
    rootViewStyle: View.propTypes.style,
    rootTextStyle: Text.propTypes.style
  };

  render() {
    const { parentViewStyle, parentTextStyle, rootViewStyle, rootTextStyle } = this.props;

    // create your stylesheet on render
    // add in the parent/root context so that relative units could be used
    // you could even set rational defaults like...
      // `parentText = Object.assign({fontSize: 16}, parentTextStyle);
      // or `rootView = Object.assign(require('./styles/root'), rootViewStyle);`
    const styles = StyleSheet.createWithContext(rawStyles, {parentViewStyle, parentTextStyle, rootViewStyle, rootTextStyle});
    return (
      <View style={styles.container}>{this.props.children}</View>
    );
  }
}
```

This way someone consuming my component could pass in the contextual styles.

**Consuming a component and passing in parent styles**
```js
import React, { StyleSheet, Component, View, Text } from 'react-native';

// import my weird component that uses ems
import MyThing from './my-thing';

// make styles normally, like a normal person
const styles = StyleSheet.create({
  view: {
    width: 100,
    height: 100,
  },
  text: {
    fontSize: 24
  }
});

export default class AnotherThing extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    // pass the parent styles to the MyThing so it can render relative units correctly
    return (
      <View style={styles.view}>
        <MyThing parentViewStyles={styles.view} parentTextStyles={styles.text} />
        <Text style={styles.text}>Hello There!</Text>
        {this.props.children}
      </View>
    );
  }
}
```

**Comment by heygrady** [comment]
You might also consider adding options like `stripUnsupportedUnits`, `baseFontSize`, `parentFontSize`, or `stripUnsupportedProperties`. Currently you leave unsupported properties in place, which is also kind of nice as far as allowing for novel extensions.

I was considering doing something like using `display: block`, `display: inline`, and `display: inline-block` to trigger "approximate" flexbox styles. Or having percentage widths converted from `width: 50%` to `flex: 0.5`

**Comment by heygrady** [comment]
BTW, I was suggesting adding a config value for `em` conversion because using a global default is better than using 1.
